Feature/Better component init

jseidl opened the pull request:
Implementing proper component initialization using `RestoreEntity` and `async_added_to_hass` to prevent Area state change upon restart of Home Assistant.

This removes entirely the need for `passive_start` which was a workaround for the lack of restore.

Closes #28

jseidl commented:
Removed `passive_start` from config and implemented restore ability on the `AreaPresenceBinarySensor`. Still have to implement on the others.

sti0 commented:
Got this error on startup:
`2020-10-26 22:30:35 ERROR (MainThread) [homeassistant.components.binary_sensor] Error while setting up magic_areas platform for binary_sensor
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 201, in _async_setup_platform
    await asyncio.gather(*pending)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 310, in async_add_entities
    await asyncio.gather(*tasks)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 481, in _async_add_entity
    await entity.add_to_platform_finish()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 521, in add_to_platform_finish
    await self.async_added_to_hass()
  File "/config/custom_components/magic_areas/binary_sensor.py", line 256, in async_added_to_hass
    _LOGGER.debug(f"Area restored: {self.slug}")
AttributeError: 'AreaPresenceBinarySensor' object has no attribute 'slug'`

jseidl commented:
That should be `self.area.slug` Gonna fix that in a bit.

sti0 commented on custom_components/magic_areas/binary_sensor.py:
```diff
@@ -253,11 +237,63 @@ def device_class(self):
+        is_new_entry = last_state is None  # newly added to HA
+
+        if is_new_entry:
+            _LOGGER.debug(f"New area detected: {self.slug}")
```
```suggestion
            _LOGGER.debug(f"New area detected: {self.area.slug}")

```

sti0 commented on custom_components/magic_areas/binary_sensor.py:
```diff
@@ -253,11 +237,63 @@ def device_class(self):
+            _LOGGER.debug(f"New area detected: {self.slug}")
+            self._update_state()
+        else:
+            _LOGGER.debug(f"Area restored: {self.slug}")
```
```suggestion
            _LOGGER.debug(f"Area restored: {self.area.slug}")

```

sti0 commented:
Bugreport:
Area is occupied and the lights are on. The `disable_entity` is off. After switching the `disable_entity` to on, the lights stay on. After switching the `disable_entity` back to off, the lights turn off.

jseidl commented:
If you turn on lights in a non-occupied area, eventually `update_interval` will kick in, see that the area is not occupied and proceed on turning the lights off. `disable_entity` and `sleep_entity` state change also trigger an area state check.

Disable to off turning lights off is weird, unless the room was already clear. Will investigate

jseidl commented:
Added `RestoreEntity` capabilities to both the Presence `binary_sensor` and the Presence Hold `switch`. The aggregators don't look like would benefit from state restore as running the averaging / grouping on start is not detrimental (at least looks like it so far).

Added hook to `EVENT_HOMEASSISTANT_STARTED` to setup all the state trackers.

Tested on my test install and seems to be working so far. Nothing on my logs and working as intended. Gonna deploy on my home install and test for a few days before merging since a lot of code was moved around on this one.

jseidl commented:
Been running this branch a couple of days. Seems to be stable. Will merge into `develop` branch this evening

sti0 commented:
> Been running this branch a couple of days. Seems to be stable. Will merge into `develop` branch this evening

I've been testing this branch today as well, but I still facing the issue mentioned in [my comment](https://github.com/jseidl/hass-magic_areas/pull/43#issuecomment-717442117) a couple of days ago.

After some testing I found the issue.
I configured a `disable_state: "off"` and it doesn't work. If I remove this configuration and it works with default state = on, just fine. So there is a bug with a custom `disable_state`.

sti0 commented on custom_components/magic_areas/binary_sensor.py:
```diff
@@ -253,24 +237,83 @@ def device_class(self):
+
+    def autolight_disable_state_change(self, entity_id, from_state, to_state):
+
         if to_state.state == STATE_OFF:
```
doesn't respect the custom `disable_state`

jseidl replied:
Good catch! Will push a fix soon

jseidl replied:
Actually this bug is on the master so should be a `hotfix/`. Will merge this one and push a hotfix to the master.

KTibow commented:
You can use draft PRs, BTW, instead of putting WIP in the title and using a bot. Draft PRs can't be merged until marked as ready for review.
